File: sdk-remote/src/swig/java/tests/timer/Timer.java
/*
 * Copyright (C) 2010, Gostai S.A.S.
 *
 * This software is provided "as is" without warranty of any kind,
 * either expressed or implied, including but not limited to the
 * implied warranties of fitness for a particular purpose.
 *
 * See the LICENSE file for more information.
 */
package tests.timer;

import urbi.Log;
import urbi.UObject;
import urbi.UVar;
import urbi.TimerHandle;

public class Timer extends UObject
{
    static {
	UStartRename(Timer.class, "timer");
    }

    public Timer(String name)
    {
	super(name);
	UBindVar(updated, "updated");
	UBindVar(timerup, "timerup");
	UBindVar(hupdated, "hupdated");
	hupdated.setValue(0);
	timerup.setValue(0);
	updated.setValue(0);
	UBindFunction("setupUpdate");
	UBindFunction("setupTimer");
	UBindFunction("unsetupTimer");
	UBindFunction("init");
    }
    public int init()
    {
	return 0;
    }
    public int setupUpdate(int d)
    {
	USetUpdate(d);
	return 0;
    }
    public String setupTimer(int d)
    {
	return USetTimer(d, this, "onTimer");
    }
    public boolean unsetupTimer(String s)
    {
	return removeTimer(TimerHandle.create(s));
    }
    public int update()
    {
	updated.setValue(updated.intValue() + 1);
	return 0;
    }
    public int onTimer()
    {
	timerup.setValue(timerup.intValue() + 1);
	return 0;
    }
    UVar updated = new UVar();
    UVar hupdated = new UVar();
    UVar timerup = new UVar();
};

File: sdk-remote/src/uobjects/test/timer.uob/timer.cc
/*
 * Copyright (C) 2008-2011, Gostai S.A.S.
 *
 * This software is provided "as is" without warranty of any kind,
 * either expressed or implied, including but not limited to the
 * implied warranties of fitness for a particular purpose.
 *
 * See the LICENSE file for more information.
 */

#include <libport/debug.hh>
#include <iostream>
#include <urbi/uobject.hh>

GD_CATEGORY(Test.Timer);

urbi::UObjectHub* thub;
class timer: public urbi::UObject
{
public:
  pthread_t mainthread;

  inline void threadCheck()
  {
    aver(mainthread == pthread_self());
  }

  timer(const std::string& name)
    : urbi::UObject(name)
  {
    mainthread = pthread_self();
    UBindVar(timer, updated);
    UBindVar(timer, timerup);
    UBindVar(timer, hupdated);
    hupdated = 0;
    timerup = 0;
    updated = 0;
    UBindFunction(timer, setupUpdate);
    UBindFunction(timer, setupTimer);
    UBindFunction(timer, unsetupTimer);
    UBindFunction(timer, setupHubUpdate);
    UBindFunction(timer, init);
  }
  int init()
  {
    threadCheck();
    return 0;
  }
  int setupUpdate(int d)
  {
    threadCheck();
    USetUpdate(d);
    return 0;
  }
  std::string setupTimer(int d)
  {
    threadCheck();
    return *USetTimer(d, &timer::onTimer);
  }
  bool unsetupTimer(const std::string& s)
  {
    threadCheck();
    return removeTimer(urbi::TimerHandle(new std::string(s)));
  }
  int setupHubUpdate(int d)
  {
    threadCheck();
    thub->USetUpdate(d);
    return 0;
  }
  virtual int update()
  {
    threadCheck();
    updated = (int)updated + 1;
    return 0;
  }
  virtual int onTimer()
  {
    threadCheck();
    timerup = (int)timerup + 1;
    return 0;
  }
  urbi::UVar updated;
  urbi::UVar hupdated;
  urbi::UVar timerup;
};

class timerHub: public urbi::UObjectHub
{
public:
  timerHub(const std::string& name)
    : urbi::UObjectHub(name)
  {
    GD_INFO_DEBUG("timerhub started");
    thub = this;
  }

  virtual int update()
  {
    urbi::UVar v("timer","hupdated");
    v = 1;
    return 0;
  }
};

UStart(timer);
UStartHub(timerHub);
